### ISSUE TITLE
unknown preference type Ú when running with tclsh8.6

### ISSUE THREAD

**Issue by ploxiln** [zdia/gorilla#193]
tclsh8.6 built locally on macOS 10.13 with homebrew, running the current latest commit of the `pre160` branch 0254c38f67c992500a5f18001ad42dc85b792cdc

This error does not occur for tclsh8.5 which is included with the base macOS.

```
Can not open password database "-errorstack {INNER {returnImm {unknown preference type: Ú} {}} CALL {my setPreferencesFromString {ÚPü?[nw¦ç 1 I 5 12 I 7 30}} CALL {::pwsafe::db0 setHeaderField 2 {ÚPü?[nw¦ç 1 I 5 12 I 7 30}} CALL {my readHeaderFields} CALL {::pwsafe::reader0 readFile pcv} CALL {pwsafe::createFromStream ::pwsafe::streamreader0 NMAtgeXM3b1+ 3 pcv} CALL {pwsafe::createFromFile /Users/pierce/current/passdb/plopez.dat NMAtgeXM3b1+ junk} CALL {OpenDatabase {Open Password Database} {} 1}} -errorcode NONE -errorinfo {unknown preference type: Ú
    while executing
"error [ mc "unknown preference type: %s" $prefType ]"
    (object "::pwsafe::db0" method "::pwsafe::db::setPreferencesFromString" body line 21)
    invoked from within
"setPreferencesFromString $value"
    (object "::pwsafe::db0" method "::pwsafe::db::setHeaderField" body line 7)
    invoked from within
"$db setHeaderField $fieldType $fieldValue"
    (object "::pwsafe::reader0" method "::pwsafe::v3::reader::readHeaderFields" body line 44)
    invoked from within
"readHeaderFields"
    (object "::pwsafe::reader0" method "::pwsafe::v3::reader::readFile" body line 123)
    invoked from within
"$reader readFile $pcvp"
    (procedure "pwsafe::createFromStream" line 17)
    invoked from within
"pwsafe::createFromStream $stream $password 3 $pcvp"
    (procedure "pwsafe::createFromFile" line 28)
    invoked from within
"pwsafe::createFromFile $fileName $password $pvar "} -errorline 177 -code 1 -level 0":
/Users/pierce/current/passdb/plopez.dat
```

**Comment by rich123** [comment]
That is interesting.  Can you report what patch level of Tcl 8.6 it is that fails.

```
Open a terminal, then type "tclsh8.6"
Once tclsh starts then type: 
set tcl_patchLevel
```
You will get back 8.6.?.  I'm looking for the ?.  Gorilla works for me on Linux with 8.6.1.  So I want to try compiling the exact point level you have, to see if it fails or not for me.

**Comment by ploxiln** [comment]
```
% set tcl_patchLevel
8.6.8
```

Thanks for taking a look :)

**Comment by rich123** [comment]
So far I've been unable to recreate this locally (on Linux).  For the version that errors out while running under 8.6 is it loading the C extensions for the crypto operations?  You can find out by running it under 8.6, then creating a fake empty new password safe file, then choosing the Help->About menu option.  At the bottom of the about window is a line: "Used C extensions: " and a possible list.

If the answer is yes, then the next item I'd request you check is to disable the C extensions (find them underneath the sources directory.  For MacOS the file extension is ".dylib".  One set is under tcllib/sha256c, the others are under "twofish".  Just rename the files to something else (i.e., change them to something like *.dylib.off".  Then restart gorilla with 8.6 and see if the password safe file opens properly.

**Comment by ploxiln** [comment]
Yes, it's as you suspect: if I create a new database, and look at About, I see some "Used C extension". If I rename all ".dylib" files to something else, then Password Gorilla works with my existing password database.

I am able to run the last released Password Gorilla ".app" build ... which bundles its own tcl/tk and just the sha256 C extension it seems.

**Comment by ploxiln** [comment]
Yes, it's due to `twofish.dylib`. I re-enabled the sha256c and stretchkey dylibs first, and all seemed to work, then enabled the twofish dylib, and ran into the original error.

**Comment by rich123** [comment]
Do you have a C compiler installed?  If yes, could you try recompiling the twofish library.  In theory it should amount to typing "make" in a command window when the current directory is the twofish directory.  If it works the result would be named twofish2tcl.so.  Just rename to .dynlib and substitute for the original inside the appropriate darwin directory (32 vs 64bit, although I believe modern MacOS is all 64 bit anymore).

**Comment by rich123** [comment]
Hmm...  That is even more odd.  I don't get the link failure (unless it is related to a much newer gcc?).  But simply making decrypt() be static inline should  not have broken anything.  Maybe solve one puzzle, only to be presented with yet another.

**Comment by ploxiln** [comment]
yeah, it's pretty recent llvm/clang, probably 4.x (but Apple branches off and versions separately, just to be confusing) (don't be confused by "4.2.1", that's the last version of actual gcc that Apple shipped, many years ago)

```
$ gcc --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 10.0.0 (clang-1000.10.44.2)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```
